src/virtio/net.c

```diff
@@ -0,0 +1,385 @@
+        /* vq is full or not initialised, drop the packet */
+        return;
+    }
+
+    /* Read the head of the descriptor chain */
+    uint16_t desc_head = virtq->avail->ring[idx % virtq->num];
+    uint16_t curr_desc_head = desc_head;
+
+    uint32_t copied = 0;
+    /* Amount of the current descriptor copied */
+    uint32_t desc_copied = 0;
+
+    struct virtio_net_hdr_mrg_rxbuf virtio_hdr = {0};
+    virtio_hdr.num_buffers = 1;
+
+    copied += copy_rx(virtq, &curr_desc_head, &desc_copied, &virtio_hdr, sizeof(struct virtio_net_hdr_mrg_rxbuf));
+    copied += copy_rx(virtq, &curr_desc_head, &desc_copied, state->rx_data + buf_offset, size);
+
+    /* Put it in the used ring */
+    virtq_enqueue_used(virtq, desc_head, copied);
+
+    /* Record that we've used this descriptor chain now */
+    vq->last_idx++;
+
+    virtio_net_respond(dev);
+}
+
+void virtio_net_handle_rx(struct virtio_net_device *state)
+{
+    struct virtio_device *dev = &state->virtio_device;
+
+    net_buff_desc_t sddf_buffer;
+    bool reprocess = true;
+
+    while (reprocess) {
+        while (net_dequeue_active(&state->rx, &sddf_buffer) != -1) {
+            /* On failure, drop packet since we don't know how long until next interrupt */
+            handle_rx_buffer(dev, sddf_buffer.io_or_offset, sddf_buffer.len);
+
+            sddf_buffer.len = 0;
+            net_enqueue_free(&state->rx, sddf_buffer);
+        }
+
+        net_request_signal_active(&state->rx);
+        reprocess = false;
+
+        if (!net_queue_empty_active(&state->rx)) {
+            net_cancel_signal_active(&state->rx);
+            reprocess = true;
+        }
+    }
+}
+
+static virtio_device_funs_t functions = {
+    .device_reset = virtio_net_reset,
+    .get_device_features = virtio_net_get_device_features,
+    .set_driver_features = virtio_net_set_driver_features,
+    .get_device_config = virtio_net_get_device_config,
+    .set_device_config = virtio_net_set_device_config,
+    .queue_notify = virtio_net_queue_notify,
+};
+
+bool virtio_mmio_net_init(struct virtio_net_device *net_dev,
+                          uint8_t mac[VIRTIO_NET_CONFIG_MAC_SZ],
+                          uintptr_t region_base,
+                          uintptr_t region_size,
+                          size_t virq,
+                          net_queue_handle_t *rx,
+                          net_queue_handle_t *tx,
+                          uintptr_t rx_data,
+                          uintptr_t tx_data,
+                          int rx_ch,
+                          int tx_ch)
+{
+    struct virtio_device *dev = &net_dev->virtio_device;
+
+    dev->data.DeviceID = DEVICE_ID_VIRTIO_NET;
+    dev->data.VendorID = VIRTIO_MMIO_DEV_VENDOR_ID;
+    dev->funs = &functions;
+    dev->vqs = net_dev->vqs;
+    dev->num_vqs = VIRTIO_NET_NUM_VIRTQ;
+    dev->virq = virq;
+    dev->device_data = net_dev;
+
+    memcpy(net_dev->config.mac, mac, VIRTIO_NET_CONFIG_MAC_SZ);
+
+    net_dev->rx = *rx;
+    net_dev->tx = *tx;
```

Comment:
Why pass pointers if you are just going to copy the struct anyways?

Reply:
Fine either way though, not really an issue.

Reply:
Its just because these queue handles are 24 bytes each and it avoids a copy. Doesn't matter though since it only impacts initialisation, its just a habbit.

Reply:
If we were using LTO the copy would probably be elided

Reply:
Yep no worries let's just keep it how it is right now.